File: 1320-remove-all-adjacent-duplicates-in-string-ii/remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) 
    {
        stack<pair<char,int>> stack;
        stack.push({s[0],1});
        for(int i=1;i<s.length();i++)
        {
            if(!stack.empty() && s[i]==stack.top().first)
            stack.top().second++;
            else
            stack.push({s[i],1});

            while(!stack.empty() && stack.top().second==k)
            stack.pop();
        }

        string ans="";
        while(!stack.empty())
        {
            char c=stack.top().first;
            int freq=stack.top().second;
            stack.pop();

            while(freq--)
            ans.push_back(c);
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: 1320-remove-all-adjacent-duplicates-in-string-ii/remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    public String removeDuplicates(String s, int k) 
    {
        // stack {char, freq}
        Deque<int[]> stack = new ArrayDeque<>();
        for(int i = 0; i < s.length(); i++)
        {
            if(!stack.isEmpty() && s.charAt(i) - 'a' == stack.peek()[0])
            stack.peek()[1]++;

            else
            stack.push(new int[]{s.charAt(i) - 'a', 1});

            if(!stack.isEmpty() && stack.peek()[1] == k)
            stack.pop();
        }

        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty())
        {
            int ch = stack.peek()[0];
            int freq = stack.poll()[1];

            while(freq > 0)
            {
                sb.append((char)(ch + 'a'));
                freq--;
            }
        }

        return sb.reverse().toString();
    }
}